File: frontend/app/lib/bracketPrograms.ts
import { BracketGroup, BracketProgramDefinition, BracketResponse, Player } from './types'

export const requiredBracketProgramKeys = ['handicap', 'scratch'] as const

export const divisionOptions = ['Men', 'Women', 'Senior', 'Junior'] as const
export type BowlerDivision = (typeof divisionOptions)[number]

export function normalizeDivision(division?: string | null): BowlerDivision {
  const value = String(division || 'Men').trim().toLowerCase()
  if (value === 'men' || value === 'mens' || value === "men's") return 'Men'
  if (value === 'womens' || value === "women's" || value === 'women') return 'Women'
  if (value === 'senior' || value === 'seniors' || value === "senior's") return 'Senior'
  if (value === 'junior' || value === 'juniors' || value === "junior's") return 'Junior'
  return 'Men'
}

export function isProgramAllowedForDivision(programDivision: string | undefined, playerDivision: string | undefined): boolean {
  const target = String(programDivision || 'Any').trim().toLowerCase()
  if (!target || target === 'any' || target === 'open') {
    return true
  }
  return normalizeDivision(target) === normalizeDivision(playerDivision)
}

export function filterEntriesForDivision(
  entries: Record<string, number> | undefined,
  programs: BracketProgramDefinition[],
  division: string | undefined,
): Record<string, number> {
  const normalized = normalizePlayerBracketEntries(entries)
  const filtered: Record<string, number> = {}

  Object.entries(normalized).forEach(([key, count]) => {
    const program = programs.find(programItem => programItem.key === key)
    if (!program || isProgramAllowedForDivision(program.division, division)) {
      filtered[key] = Math.max(0, Number(count || 0))
    }
  })

  return filtered
}

const legacyBracketProgramKeyMap: Record<string, string> = {
  womens: 'womens_scratch',
  seniors: 'seniors_scratch',
  juniors: 'juniors_scratch',
}

function canonicalizeBracketProgramKey(key: string): string {
  return legacyBracketProgramKeyMap[key] ?? key
}

export const defaultBracketPrograms: BracketProgramDefinition[] = [
  {
    key: 'handicap',
    name: 'Handicap',
    division: 'Any',
    scoring_mode: 'handicap',
    enabled: true,
    allow_byes: false,
    display_order: 1,
  },
  {
    key: 'scratch',
    name: 'Scratch',
    division: 'Any',
    scoring_mode: 'scratch',
    enabled: true,
    allow_byes: false,
    display_order: 2,
  },
  {
    key: 'reverse_scratch',
    name: 'Reverse Scratch',
    division: 'Any',
    scoring_mode: 'reverse_scratch',
    enabled: false,
    allow_byes: false,
    display_order: 3,
  },
  {
    key: 'reverse_handicap',
    name: 'Reverse Handicap',
    division: 'Any',
    scoring_mode: 'reverse_handicap',
    enabled: false,
    allow_byes: false,
    display_order: 4,
  },
  {
    key: 'womens_scratch',
    name: "Women's Scratch",
    division: 'Womens',
    scoring_mode: 'scratch',
    enabled: false,
    allow_byes: false,
    display_order: 5,
  },
  {
    key: 'womens_handicap',
    name: "Women's Handicap",
    division: 'Womens',
    scoring_mode: 'handicap',
    enabled: false,
    allow_byes: false,
    display_order: 6,
  },
  {
    key: 'seniors_scratch',
    name: 'Seniors Scratch',
    division: 'Senior',
    scoring_mode: 'scratch',
    enabled: false,
    allow_byes: false,
    display_order: 7,
  },
  {
    key: 'seniors_handicap',
    name: 'Seniors Handicap',
    division: 'Senior',
    scoring_mode: 'handicap',
    enabled: false,
    allow_byes: false,
    display_order: 8,
  },
  {
    key: 'juniors_scratch',
    name: 'Juniors Scratch',
    division: 'Junior',
    scoring_mode: 'scratch',
    enabled: false,
    allow_byes: false,
    display_order: 9,
  },
  {
    key: 'juniors_handicap',
    name: 'Juniors Handicap',
    division: 'Junior',
    scoring_mode: 'handicap',
    enabled: false,
    allow_byes: false,
    display_order: 10,
  },
]

function getDefaultBracketProgram(key: string): BracketProgramDefinition | undefined {
  return defaultBracketPrograms.find(program => program.key === key)
}

function shouldUseCanonicalProgramName(key: string, name: string | undefined): boolean {
  const normalizedName = (name || '').trim().toLowerCase()
  return (
    !normalizedName
    || (key === 'womens_scratch' && normalizedName === 'womens')
    || (key === 'seniors_scratch' && normalizedName === 'seniors')
    || (key === 'juniors_scratch' && normalizedName === 'juniors')
  )
}

export function normalizeBracketPrograms(
  programs: BracketProgramDefinition[] | undefined,
  fallbackEntryFee?: number,
): BracketProgramDefinition[] {
  const source = programs?.length ? [...programs] : []
  const configuredKeys = new Set(
    source
      .map(program => canonicalizeBracketProgramKey(String(program.key || '').trim().toLowerCase().replace(/\s+/g, '-')))
      .filter(Boolean),
  )

  defaultBracketPrograms.forEach(program => {
    if (!configuredKeys.has(program.key)) {
      source.push(program)
    }
  })

  const seen = new Set<string>()

  const normalized = source
    .map((program, index) => {
      const rawKey = String(program.key || '')
        .trim()
        .toLowerCase()
        .replace(/\s+/g, '-')
      const key = canonicalizeBracketProgramKey(rawKey)
      if (!key || seen.has(key)) {
        return null
      }
      seen.add(key)
      const canonicalProgramDefaults = getDefaultBracketProgram(key)
      const aliasedProgramDefaults = rawKey !== key ? canonicalProgramDefaults : undefined
      return {
        ...aliasedProgramDefaults,
        ...program,
        key,
        name: (
          (shouldUseCanonicalProgramName(key, program.name) ? canonicalProgramDefaults?.name : undefined)
          || aliasedProgramDefaults?.name
          || program.name
          || key.replace(/-/g, ' ').replace(/\b\w/g, letter => letter.toUpperCase())
        ).trim(),
        division: (canonicalProgramDefaults?.division || aliasedProgramDefaults?.division || program.division || 'Any').trim() || 'Any',
        scoring_mode: (canonicalProgramDefaults?.scoring_mode || aliasedProgramDefaults?.scoring_mode || program.scoring_mode || key).trim().toLowerCase(),
        entry_fee: (program.entry_fee != null && program.entry_fee > 0) ? program.entry_fee : fallbackEntryFee,
        enabled: requiredBracketProgramKeys.includes(key as (typeof requiredBracketProgramKeys)[number]) ? true : (program.enabled ?? false),
        allow_byes: Boolean(program.allow_byes ?? canonicalProgramDefaults?.allow_byes ?? aliasedProgramDefaults?.allow_byes ?? false),
        display_order: program.display_order ?? canonicalProgramDefaults?.display_order ?? aliasedProgramDefaults?.display_order ?? (index + 1),
      }
    })
    .filter(Boolean) as BracketProgramDefinition[];
  return normalized.sort((left, right) => (left.display_order ?? 0) - (right.display_order ?? 0))
}

export function getEnabledBracketPrograms(programs: BracketProgramDefinition[] | undefined): BracketProgramDefinition[] {
  return normalizeBracketPrograms(programs).filter(program =>
    (program.enabled ?? false) || requiredBracketProgramKeys.includes(program.key as (typeof requiredBracketProgramKeys)[number])
  )
}

export function normalizePlayerBracketEntries(
  bracketEntries: Record<string, number> | undefined,
  handicap?: number,
  scratch?: number,
): Record<string, number> {
  const normalized: Record<string, number> = {}

  Object.entries(bracketEntries || {}).forEach(([key, value]) => {
    const normalizedKey = canonicalizeBracketProgramKey(key.trim().toLowerCase().replace(/\s+/g, '-'))
    if (!normalizedKey) return
    normalized[normalizedKey] = Math.max(0, Number(value || 0))
  })

  if (normalized.handicap === undefined && handicap) {
    normalized.handicap = Math.max(0, handicap)
  }
  if (normalized.scratch === undefined && scratch) {
    normalized.scratch = Math.max(0, scratch)
  }

  return normalized
}

export function calculatePlayerTotalCost(
  bracketEntries: Record<string, number> | undefined,
  programs: BracketProgramDefinition[],
  fallbackEntryFee: number,
): number {
  const programMap = new Map(programs.map(program => [program.key, program]))
  const normalizedEntries = normalizePlayerBracketEntries(bracketEntries)

  return Object.entries(normalizedEntries).reduce((total, [key, count]) => {
    const progFee = programMap.get(key)?.entry_fee
    const fee = Number((progFee != null && progFee > 0) ? progFee : (fallbackEntryFee ?? 0))
    return total + (Math.max(0, count) * fee)
  }, 0)
}

export function getBracketGroups(response: BracketResponse | null | undefined): BracketGroup[] {
  if (!response) return []
  if (response.bracket_groups?.length) {
    return response.bracket_groups
  }

  const groups: BracketGroup[] = []
  if (response.scratch_brackets?.length) {
    groups.push({ key: 'scratch', name: 'Scratch', scoring_mode: 'scratch', brackets: response.scratch_brackets })
  }
  if (response.handicap_brackets?.length) {
    groups.push({ key: 'handicap', name: 'Handicap', scoring_mode: 'handicap', brackets: response.handicap_brackets })
  }
  if (response.multiple_brackets?.scratch_brackets?.length) {
    groups.push({ key: 'scratch', name: 'Scratch', scoring_mode: 'scratch', brackets: response.multiple_brackets.scratch_brackets })
  }
  if (response.multiple_brackets?.handicap_brackets?.length) {
    groups.push({ key: 'handicap', name: 'Handicap', scoring_mode: 'handicap', brackets: response.multiple_brackets.handicap_brackets })
  }
  return groups
}

export function summarizeEntries(players: Player[], programs: BracketProgramDefinition[], bracketSize: number, fallbackEntryFee: number) {
  const programSummaries = programs.map(program => {
    const counts = players.map(player =>
      (player as unknown as { bracketEntries?: Record<string, number> }).bracketEntries?.[program.key]
      ?? player.programEntryCounts?.[program.key]
      ?? 0
    )
    const totalEntries = counts.reduce((sum, count) => sum + count, 0)
    const fillResult = simulateBracketFill(counts, bracketSize)

    return {
      ...program,
      totalEntries,
      expectedBrackets: fillResult.brackets,
      refunds: fillResult.refunds,
      revenue: totalEntries * Number(program.entry_fee ?? fallbackEntryFee ?? 0),
    }
  })

  const totalEntries = programSummaries.reduce((sum, program) => sum + program.totalEntries, 0)
  const totalRevenue = players.reduce((sum, player) => {
    const isPaid = player.amountPaid >= player.totalCost
    return sum + (isPaid ? player.totalCost : 0)
  }, 0)

  return {
    totalPlayers: players.length,
    totalEntries,
    totalRevenue,
    programSummaries,
  }
}

function simulateBracketFill(playerCounts: number[], bracketSize: number) {
  const total = playerCounts.reduce((sum, count) => sum + count, 0)
  if (total < bracketSize) {
    return { brackets: 0, refunds: total }
  }

  let bracketCount = Math.floor(total / bracketSize)
  while (bracketCount > 0) {
    const fillable = playerCounts.reduce((sum, count) => sum + Math.min(count, bracketCount), 0)
    const nextBracketCount = Math.floor(fillable / bracketSize)
    if (nextBracketCount >= bracketCount) {
      break
    }
    bracketCount = nextBracketCount
  }

  return {
    brackets: bracketCount,
    refunds: total - (bracketCount * bracketSize),
  }
}

File: backend/app/core/bracket_programs.py
from __future__ import annotations

from decimal import Decimal
from typing import Any, Dict, Iterable, List, Mapping


DEFAULT_BRACKET_PROGRAMS: List[Dict[str, Any]] = [
    {
        "key": "handicap",
        "name": "Handicap",
        "division": "Any",
        "scoring_mode": "handicap",
        "entry_fee": None,
        "enabled": True,
        "allow_byes": False,
        "display_order": 1,
    },
    {
        "key": "scratch",
        "name": "Scratch",
        "division": "Any",
        "scoring_mode": "scratch",
        "entry_fee": None,
        "enabled": True,
        "allow_byes": False,
        "display_order": 2,
    },
    {
        "key": "reverse_scratch",
        "name": "Reverse Scratch",
        "division": "Any",
        "scoring_mode": "reverse_scratch",
        "entry_fee": None,
        "enabled": False,
        "allow_byes": False,
        "display_order": 3,
    },
    {
        "key": "reverse_handicap",
        "name": "Reverse Handicap",
        "division": "Any",
        "scoring_mode": "reverse_handicap",
        "entry_fee": None,
        "enabled": False,
        "allow_byes": False,
        "display_order": 4,
    },
    {
        "key": "womens_scratch",
        "name": "Women's Scratch",
        "division": "Womens",
        "scoring_mode": "scratch",
        "entry_fee": None,
        "enabled": False,
        "allow_byes": False,
        "display_order": 5,
    },
    {
        "key": "womens_handicap",
        "name": "Women's Handicap",
        "division": "Womens",
        "scoring_mode": "handicap",
        "entry_fee": None,
        "enabled": False,
        "allow_byes": False,
        "display_order": 6,
    },
    {
        "key": "seniors_scratch",
        "name": "Seniors Scratch",
        "division": "Senior",
        "scoring_mode": "scratch",
        "entry_fee": None,
        "enabled": False,
        "allow_byes": False,
        "display_order": 7,
    },
    {
        "key": "seniors_handicap",
        "name": "Seniors Handicap",
        "division": "Senior",
        "scoring_mode": "handicap",
        "entry_fee": None,
        "enabled": False,
        "allow_byes": False,
        "display_order": 8,
    },
    {
        "key": "juniors_scratch",
        "name": "Juniors Scratch",
        "division": "Junior",
        "scoring_mode": "scratch",
        "entry_fee": None,
        "enabled": False,
        "allow_byes": False,
        "display_order": 9,
    },
    {
        "key": "juniors_handicap",
        "name": "Juniors Handicap",
        "division": "Junior",
        "scoring_mode": "handicap",
        "entry_fee": None,
        "enabled": False,
        "allow_byes": False,
        "display_order": 10,
    },
]

REQUIRED_BRACKET_PROGRAM_KEYS = {"handicap", "scratch"}
LEGACY_BRACKET_PROGRAM_KEY_MAP = {
    "womens": "womens_scratch",
    "seniors": "seniors_scratch",
    "juniors": "juniors_scratch",
}

LEGACY_BRACKET_PROGRAM_NAME_MAP = {
    "womens_scratch": "womens",
    "seniors_scratch": "seniors",
    "juniors_scratch": "juniors",
}


def _canonicalize_bracket_program_key(key: str) -> str:
    return LEGACY_BRACKET_PROGRAM_KEY_MAP.get(key, key)


DEFAULT_BRACKET_PROGRAM_BY_KEY = {
    program["key"]: program for program in DEFAULT_BRACKET_PROGRAMS
}


def normalize_bracket_programs(
    bracket_programs: Iterable[Mapping[str, Any]] | None,
    default_entry_fee: float | None = None,
) -> List[Dict[str, Any]]:
    programs = list(bracket_programs or [])
    configured_keys = {
        _canonicalize_bracket_program_key(
            str(program.get("key") or "").strip().lower().replace(" ", "-")
        )
        for program in programs
        if str(program.get("key") or "").strip()
    }

    for default_program in DEFAULT_BRACKET_PROGRAMS:
        if default_program["key"] not in configured_keys:
            programs.append(default_program)

    normalized: List[Dict[str, Any]] = []
    seen_keys: set[str] = set()

    for index, program in enumerate(programs):
        source_key = str(program.get("key") or "").strip().lower().replace(" ", "-")
        raw_key = _canonicalize_bracket_program_key(source_key)
        if not raw_key or raw_key in seen_keys:
            continue

        seen_keys.add(raw_key)
        aliased_defaults = DEFAULT_BRACKET_PROGRAM_BY_KEY.get(raw_key) if source_key != raw_key else None
        normalized_name = str(program.get("name") or "").strip().lower()
        legacy_name = LEGACY_BRACKET_PROGRAM_NAME_MAP.get(raw_key)
        should_use_canonical_name = (
            not normalized_name
            or (legacy_name is not None and normalized_name == legacy_name)
        )
        normalized.append(
            {
                "key": raw_key,
                "name": str(
                    (DEFAULT_BRACKET_PROGRAM_BY_KEY.get(raw_key, {}) if should_use_canonical_name else {}).get("name")
                    or (aliased_defaults or {}).get("name")
                    or program.get("name")
                    or raw_key.replace("-", " ").title()
                ).strip(),
                "division": str(
                    (aliased_defaults or {}).get("division")
                    or program.get("division")
                    or "Any"
                ).strip()
                or "Any",
                "scoring_mode": str(
                    (aliased_defaults or {}).get("scoring_mode")
                    or program.get("scoring_mode")
                    or raw_key
                ).strip().lower()
                or raw_key,
                "entry_fee": _coerce_float(program.get("entry_fee"), default_entry_fee),
                "enabled": True if raw_key in REQUIRED_BRACKET_PROGRAM_KEYS else bool(program.get("enabled", False)),
                "allow_byes": bool(
                    program.get("allow_byes")
                    if program.get("allow_byes") is not None
                    else (aliased_defaults or {}).get("allow_byes", False)
                ),
                "display_order": int(
                    program.get("display_order")
                    or (aliased_defaults or {}).get("display_order")
                    or (index + 1)
                ),
            }
        )

    normalized.sort(key=lambda item: (item["display_order"], item["name"].lower()))
    return normalized


def normalize_bowler_bracket_entries(
    bracket_entries: Mapping[str, Any] | None,
    handicap_entries: int | None = None,
    scratch_entries: int | None = None,
) -> Dict[str, int]:
    normalized: Dict[str, int] = {}

    for raw_key, raw_value in (bracket_entries or {}).items():
        key = _canonicalize_bracket_program_key(
            str(raw_key).strip().lower().replace(" ", "-")
        )
        if not key:
            continue
        normalized[key] = max(0, int(raw_value or 0))

    if "handicap" not in normalized and handicap_entries:
        normalized["handicap"] = max(0, int(handicap_entries or 0))
    if "scratch" not in normalized and scratch_entries:
        normalized["scratch"] = max(0, int(scratch_entries or 0))

    return normalized


def calculate_bowler_total_cost(
    bracket_entries: Mapping[str, Any] | None,
    bracket_programs: Iterable[Mapping[str, Any]] | None,
    default_entry_fee: float | None,
    handicap_entries: int | None = None,
    scratch_entries: int | None = None,
) -> float:
    normalized_entries = normalize_bowler_bracket_entries(
        bracket_entries,
        handicap_entries=handicap_entries,
        scratch_entries=scratch_entries,
    )
    normalized_programs = normalize_bracket_programs(
        bracket_programs,
        default_entry_fee=default_entry_fee,
    )
    program_map = {program["key"]: program for program in normalized_programs}

    total = Decimal("0")
    for key, count in normalized_entries.items():
        if count <= 0:
            continue
        entry_fee = _coerce_decimal(program_map.get(key, {}).get("entry_fee"), default_entry_fee)
        total += entry_fee * count

    return float(total)


def _coerce_float(value: Any, fallback: float | None) -> float | None:
    if value is None or value == "":
        return fallback
    return float(value)


def _coerce_decimal(value: Any, fallback: float | None) -> Decimal:
    if value is None or value == "":
        value = fallback or 0
    return Decimal(str(value))